### PR TITLE
feat: Add REST api support

### PR DESCRIPTION
**Pull request by Astalaseven**
(for review only, don't merge yet)

### CODE REVIEW

File: app/src/main/java/gitlin/kothub/github/api/Base.kt [hunk 1]
```diff
@@ -5,21 +5,46 @@ import com.github.kittinunf.fuel.android.extension.responseJson
...
         .body("{ \"query\": \"$query\" }")
         .responseJson { request, response, result ->
             when (result) {
                 is Result.Failure -> callback(result.error, null)
                 is Result.Success -> callback(null, result.value.obj().obj("data")!!)
             }
         }
+}
+
+fun getObj(query: String, callback: (FuelError?, JSONObject?) -> Unit): Request {
```
Comment (fknop) [inline app/src/main/java/gitlin/kothub/github/api/Base.kt]: getObj and getArr are almost the same functions

File: app/src/main/java/gitlin/kothub/github/api/Viewer.kt [hunk 2]
```diff
@@ -92,4 +90,20 @@ fun drawerInfo (callback: (FuelError?, DrawerInfo?) -> Unit) {
...
+fun feeds (callback: (FuelError?, Feeds?) -> Unit) {
+    getObj("feeds", {
+        error, result ->
+
+        callback(error, if (result == null) null else Feeds(result))
+    } )
+}
+
+fun notifications (callback: (FuelError?, Notifications?) -> Unit) {
+    getArr("notifications", {
```
Comment (fknop) [inline app/src/main/java/gitlin/kothub/github/api/Viewer.kt]: You can close the parenthesis and add the lambda after without the comma.

File: app/src/main/java/gitlin/kothub/github/api/Viewer.kt [hunk 3]
```diff
@@ -92,4 +90,20 @@ fun drawerInfo (callback: (FuelError?, DrawerInfo?) -> Unit) {
 
         callback(error, if (result == null) null else DrawerInfo(result))
     }
+}
+
+fun feeds (callback: (FuelError?, Feeds?) -> Unit) {
+    getObj("feeds", {
```
Comment (fknop) [inline app/src/main/java/gitlin/kothub/github/api/Viewer.kt]: You can close the parenthesis and add the lambda after without the comma.

File: app/src/main/java/gitlin/kothub/github/api/data/Notifications.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+import gitlin.kothub.utilities.getValue
+import gitlin.kothub.utilities.obj
+import org.json.JSONObject
+import org.json.JSONArray
+
+import gitlin.kothub.utilities.*
+
+data class Notifications(private val json: JSONArray) {
+    val x = json.map<JSONObject, Notification> { Notification(it) }
```
Comment (fknop) [inline app/src/main/java/gitlin/kothub/github/api/data/Notifications.kt]: What is x?

File: app/src/main/java/gitlin/kothub/utilities/JsonUtilities.kt [hunk 5]
```diff
@@ -23,4 +23,7 @@ fun <T, V> JSONArray.map(mapper: (T) -> V): MutableList<V> {
     return (0..length() - 1).mapTo(arrayListOf<V>()) { mapper(get(it) as T) }
 }
 
+operator fun JSONArray.iterator(): Iterator<JSONObject>
```
Comment (fknop) [inline app/src/main/java/gitlin/kothub/utilities/JsonUtilities.kt]: Why is this needed ? Maybe a foreach instead ?

**Review by fknop — CHANGES_REQUESTED** [review]

File: app/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -23,6 +23,7 @@
         <activity android:name=".github.LoginActivity" />
         <activity android:name=".github.LoginWebViewActivity" />
         <activity android:name=".ProfileActivity" />
+        <activity android:name=".NotificationActivity" />
```
Comment (fknop) [inline app/src/main/AndroidManifest.xml]: Move NotificationActivity to ui package

File: app/src/main/java/gitlin/kothub/MainActivity.kt [hunk 7]
```diff
@@ -40,7 +40,7 @@ class MainActivity : AppCompatActivity(), AnkoLogger {
 
         debug(OAuthValues.isLoggedIn)
         if (OAuthValues.isLoggedIn) {
-            startActivity(Intent(this, ProfileActivity::class.java))
+            startActivity(Intent(this, NotificationActivity::class.java))
```
Comment (fknop) [inline app/src/main/java/gitlin/kothub/MainActivity.kt]: Add drawer item instead

File: app/src/main/java/gitlin/kothub/NotificationActivity.kt [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+            }
+        }
+
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_notifs)
+        setSupportActionBar(toolbar)
+
+        //AppDrawer(this, toolbar)
```
Comment (fknop) [inline app/src/main/java/gitlin/kothub/NotificationActivity.kt]: Add the drawer and select the notifications item after adding it to the drawer

File: app/src/main/java/gitlin/kothub/github/api/data/Notifications.kt [hunk 9]
```diff
@@ -1,42 +1,48 @@
...
+import com.google.gson.*
 
 import gitlin.kothub.utilities.*
 
-data class Notifications(private val json: JSONArray) {
-    val x = json.map<JSONObject, Notification> { Notification(it) }
+private val gson = Gson()
+
+class Notifications(private val json: String) {
+    private val arr: JsonArray = JsonParser().parse(json).asJsonArray
```
Comment (fknop) [inline app/src/main/java/gitlin/kothub/github/api/data/Notifications.kt]: Would be better to parse it and pass the json array to the notifications object

**Review by fknop — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/activity_notifs.xml [hunk 10]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:id="@+id/listLayout"
+    android:orientation="vertical" android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
```
Comment (fknop) [inline app/src/main/res/layout/activity_notifs.xml:6]: You need to include the toolbar xml file here

**Review by fknop — CHANGES_REQUESTED** [review]

File: app/src/main/java/gitlin/kothub/utilities/JsonUtilities.kt (hunk 5 above)
Comment (Astalaseven) [reply app/src/main/java/gitlin/kothub/utilities/JsonUtilities.kt]: It's not needed anymore

File: app/src/main/res/layout/activity_notifs.xml (hunk 10 above)
Comment (Astalaseven) [reply app/src/main/res/layout/activity_notifs.xml:6]: Done https://github.com/kothub/kothub/pull/13/commits/475d9f4987e5f512071b9189b3630bc00d8f6a25#diff-ca04348500ca962186d699cd2c441654R7

File: app/src/main/java/gitlin/kothub/github/api/data/Notifications.kt (hunk 9 above)
Comment (Astalaseven) [reply app/src/main/java/gitlin/kothub/github/api/data/Notifications.kt]: Done https://github.com/kothub/kothub/pull/13/commits/475d9f4987e5f512071b9189b3630bc00d8f6a25#diff-1d7244653ed86f90bc94cfff42c6dbccR100

File: app/src/main/java/gitlin/kothub/MainActivity.kt (hunk 7 above)
Comment (Astalaseven) [reply app/src/main/java/gitlin/kothub/MainActivity.kt]: Done https://github.com/kothub/kothub/pull/13/commits/475d9f4987e5f512071b9189b3630bc00d8f6a25#diff-eb219dfe15e958903f27572f8a62af1eR63

File: app/src/main/java/gitlin/kothub/utilities/JsonUtilities.kt (hunk 5 above)
Comment (Astalaseven) [reply app/src/main/java/gitlin/kothub/utilities/JsonUtilities.kt]: Removed https://github.com/kothub/kothub/pull/13/commits/475d9f4987e5f512071b9189b3630bc00d8f6a25#diff-46b3f60c9247b0f501484437df1a2745L29

File: app/src/main/java/gitlin/kothub/NotificationActivity.kt (hunk 8 above)
Comment (Astalaseven) [reply app/src/main/java/gitlin/kothub/NotificationActivity.kt]: Drawer added https://github.com/kothub/kothub/pull/13/commits/475d9f4987e5f512071b9189b3630bc00d8f6a25#diff-b3d1cb29368e1520fbaf99b2fa2b99e6R41
Need to merge https://github.com/kothub/kothub/pull/14 first to add it to the drawer

File: app/src/main/java/gitlin/kothub/ui/Drawer.kt [hunk 11]
```diff
@@ -119,11 +127,12 @@ class AppDrawer(private val activity: AppCompatActivity, toolbar: Toolbar): Anko
             .withActionBarDrawerToggle(true)
             .withCloseOnClick(true)
             .withDelayDrawerClickEvent(250)
-            .addDrawerItems(feed, issues, pulls, DividerDrawerItem(), settings, DividerDrawerItem(), rate)
+            .addDrawerItems(feed, issues, pulls, notifs, DividerDrawerItem(), settings, DividerDrawerItem(), rate)
```
Comment (fknop) [inline app/src/main/java/gitlin/kothub/ui/Drawer.kt]: Maybe add a divider ?

**Review by fknop — APPROVED** [review]

**Pull request merged**